plugins/SocketPlugin.js

```diff
@@ -65,88 +64,169 @@ function SocketPlugin() {
         _signalWriteSemaphore() { this._signalSemaphore(this.writeSemaIndex); },
 
+        _otherEndClosed() {
+          this.status = plugin.Socket_OtherEndClosed;
+          this._signalConnSemaphore();
+        },
+
         _hostAndPort() { return this.host + ':' + this.port; },
 
         _requestNeedsProxy() {
           return plugin.needProxy.has(this._hostAndPort());
         },
 
-        _getURL(targetURL, httpMethod, isRetry) {
-          var url = '';
-
-          if (isRetry || this._requestNeedsProxy()) {
+        _getProxy(httpMethod) {
+          if (httpMethod === 'GET') {
             // crossorigin.me is preferred but does not support PUT/POST
-            if (httpMethod === 'GET') {
-              url += 'https://crossorigin.me/';
-            } else {
-              url += 'https://cors-anywhere.herokuapp.com/';
-            }
+            return 'https://crossorigin.me/';
           }
+          return 'https://cors-anywhere.herokuapp.com/';
+        },
 
+        _getURL(targetURL) {
+          var url = '';
           if (this.port !== 443) {
             url += 'http://' + this._hostAndPort() + targetURL;
           } else {
             url += 'https://' + this.host + targetURL;
           }
+          return url;
+        },
 
+        _getFullURL(targetURL, httpMethod, isRetry) {
+          var url = '';
+          if (isRetry || this._requestNeedsProxy()) {
+            url += this._getProxy(httpMethod);
+          }
+          url += this._getURL(targetURL);
           return url;
         },
 
         _performRequest() {
           var request = new TextDecoder("utf-8").decode(this.sendBuffer);
-          var requestLines = request.split('\n');
+          var headerLines = request.split('\r\n\r\n')[0].split('\n');
           // Split header lines and parse first line
-          var firstLineItems = request.split('\n')[0].split(' ');
-          var httpMethod = firstLineItems[0];
+          var firstHeaderLineItems = headerLines[0].split(' ');
+          var httpMethod = firstHeaderLineItems[0];
           if (httpMethod !== 'GET' && httpMethod !== 'PUT' &&
               httpMethod !== 'POST') {
-            this.status = plugin.Socket_OtherEndClosed;
-            this._signalConnSemaphore();
+            this._otherEndClosed();
             return -1;
           }
-          var targetURL = firstLineItems[1];
+          var targetURL = firstHeaderLineItems[1];
+
+          // Extract possible data to send
+          var data = null;
+          for (var i = 1; i < headerLines.length; i++) {
+            var line = headerLines[i];
+            if (line.indexOf('Content-Length: ') === 0) {
+              var contentLength = parseInt(line.substr(16));
+              var end = this.sendBuffer.byteLength;
+              data = this.sendBuffer.slice(end - contentLength, end);
+              break;
+            }
+          }
+
+          if (self.fetch) {
```

Comment:
That'd be `window` and `self.fetch` is [apparently the way to check](https://developer.mozilla.org/en-US/docs/Web/API/Fetch_API/Using_Fetch#Feature_detection) for fetch API support.